smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/PaginatorGenerator.kt

```diff
@@ -0,0 +1,217 @@
+                    ClientRuntimeTypes.Core.PaginatorSequence,
+                    inputSymbol,
+                    outputSymbol,
+                    inputSymbol,
+                    outputSymbol,
+                    operationShape.camelCaseName()
+                )
+            }
+        }
+
+        writer.write("")
+
+        writer.openBlock("extension \$N: \$N {", "}", inputSymbol, ClientRuntimeTypes.Core.PaginateToken) {
+            writer.openBlock("public func usingPaginationToken(_ token: \$N) -> \$N {", "}", SwiftTypes.String, inputSymbol) {
+                writer.writeInline("return ")
+                    .call {
+                        val inputShape = model.expectShape(operationShape.input.get())
+                        writer.writeInline("\$N(", inputSymbol)
+                            .indent()
+                            .call {
+                                val sortedMembers = inputShape.members().sortedBy { it.camelCaseName() }
+                                for ((index, member) in sortedMembers.withIndex()) {
+                                    if (member.memberName != markerLiteral) {
+                                        writer.writeInline("\n\$L: \$L", member.camelCaseName(), "self.${member.camelCaseName()}")
+                                    } else {
+                                        writer.writeInline("\n\$L: \$L", member.camelCaseName(), "token")
+                                    }
+                                    if (index < sortedMembers.size - 1) {
+                                        writer.writeInline(",")
+                                    }
+                                }
+                            }
+                            .dedent()
+                            .writeInline("\n)")
+                    }
+            }
+        }
+    }
+
+    // Generate a paginator that iterates over the model-specified item
+    private fun renderItemPaginator(
+        writer: SwiftWriter,
+        serviceShape: ServiceShape,
+        operationShape: OperationShape,
+        itemDesc: ItemDescriptor,
+        inputSymbol: Symbol,
+        outputSymbol: Symbol,
+    ) {
+        writer.write("")
+        val docBody = """
+        This paginator transforms the `AsyncSequence` returned by `${operationShape.camelCaseName()}Paginated`
+        to access the nested member `${itemDesc.itemSymbol.fullName}`
+        - Returns: `${itemDesc.itemSymbol.fullName}`
+        """.trimIndent()
+
+        writer.writeSingleLineDocs {
+            this.write(docBody)
+        }
+
+        writer.openBlock("extension PaginatorSequence where Input == \$N, Output == \$N {", "}", inputSymbol, outputSymbol) {
+            writer.openBlock("func \$L() async throws -> \$N {", "}", itemDesc.itemLiteral, itemDesc.itemSymbol) {
+                writer.write("return try await self.asyncCompactMap { item in item.\$L }", itemDesc.itemPathLiteral)
```

Comment:
so does this always read to the end? (as in, it forces loading the entire stream into memory, no way to read it lazily)

Reply:
ya I couldnt figure out how to return another `AsyncSequence` of that nested type. the higher level order functions like .map, or .forEach don't seem to cover this use case right now since this type is new to the language. I was thinking this is a temp situation until we can properly transform it and turn it into an `AsyncSequence` of the nested type so it's still lazy. unless you have a better suggestion? I tried like a million different thigns and couldn't get any of them to compile.